salvo/src/lib/benchmark/base_benchmark.py

```diff
@@ -78,20 +91,54 @@ def run_image(self, image_name : str, **kwargs) -> bytearray:
           here https://docker-py.readthedocs.io/en/stable/containers.html#docker.models.containers.ContainerCollection.run
 
     Returns:
-        A bytearray containing the output produced from executing the specified container
+        A bytearray containing the output produced from executing the specified
+        container
     """
-    pass
+    return self._docker_image.run_image(image_name, **kwargs)
 
   def pull_images(self) -> List[str]:
     """Retrieve all images necessary for the benchmark.
 
-    Retrieve the NightHawk and Envoy images defined in the control object. 
+    Retrieve the NightHawk and Envoy images defined in the control
+    object.
     """
-    pass
+    retrieved_images = []
+    images = self.get_images()
+
+    for image in [
+        images.nighthawk_benchmark_image,
+        images.nighthawk_binary_image,
+        images.envoy_image
+    ]:
+      # If the image name is not defined, we will have an empty string.
+      # For unit testing we'll keep this behavior. For true usage, we
+      # should raise an exception when the benchmark class performs its
+      # validation
+      if image:
+        i = self._docker_image.pull_image(image)
+        log.debug(f"Retrieved image: {i} for {image}")
+        if i is None:
+          return []
```

Comment:
Looks like we bail out after the first failure - just making sure that is intentional.

Reply:
Yes.  We aren't using this method yet, outside of tests, so we can adjust this behavior later if necessary.